api/catalog/api/views/media_views.py

```diff
@@ -143,24 +157,69 @@ def _get_user_ip(request):
         return ip
 
     @staticmethod
-    def _get_proxied_image(image_url, width=settings.THUMBNAIL_WIDTH_PX):
-        if width is None:  # full size
-            proxy_upstream = f"{settings.THUMBNAIL_PROXY_URL}/{image_url}"
-        else:
-            proxy_upstream = (
-                f"{settings.THUMBNAIL_PROXY_URL}/"
-                f"{settings.THUMBNAIL_WIDTH_PX},fit/"
-                f"{image_url}"
-            )
+    def _thumbnail_proxy_comm(
+        path: str,
+        params: dict,
+        headers: List[
+            tuple[str, str]
+        ] = None,  # ``List`` because there is a ``list`` function
+    ):
+        proxy_url = settings.THUMBNAIL_PROXY_URL
+        query_string = urlencode(params)
+        upstream_url = f"{proxy_url}/{path}?{query_string}"
+        log.debug(f"Image proxy upstream URL: {upstream_url}")
+
         try:
-            upstream_response = urlopen(proxy_upstream)
-            status = upstream_response.status
+            req = Request(upstream_url)
+            if headers:
+                for key, val in headers:
+                    req.add_header(key, val)
+            upstream_response = urlopen(req)
+
+            res_status = upstream_response.status
             content_type = upstream_response.headers.get("Content-Type")
+            log.debug(
+                "Image proxy response "
+                f"status: {res_status}, content-type: {content_type}"
+            )
+
+            return upstream_response, res_status, content_type
         except HTTPError:
             raise get_api_exception("Failed to render thumbnail.")
 
+    @staticmethod
+    def _get_proxied_image(
+        image_url: str,
+        accept_header: str = "image/*",
+        is_full_size: bool = False,
+        is_compressed: bool = True,
+    ):
+        info_res, *_ = MediaViewSet._thumbnail_proxy_comm("info", {"url": image_url})
+        info = json.loads(info_res.read())
+
+        params = {
+            "url": image_url,
+            "width": info["width"] if is_full_size else settings.THUMBNAIL_WIDTH_PX,
+        }
+
+        if is_compressed:
+            params |= {
+                "quality": settings.THUMBNAIL_JPG_QUALITY,
+                "compression": settings.THUMBNAIL_PNG_COMPRESSION,
+            }
+        else:
+            params |= {
+                "quality": 100,
+                "compression": 0,
+            }
+
+        if "webp" in accept_header:
+            params["type"] = "auto"  # Use ``Accept`` header to determine output type.
```

Comment:
This converts JPG and PNG to WEBP if the `Accept` header contains it. Otherwise the format stays as it is. This is to prevent issues with Safari which before Big Sur, [preferentially requests PNG](https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values). JPG images become quite large if converted to PNG.